File: frontend/js/views/cash_flow_statement.js
// Cash flow statement (indirect method) generator — derive CFO/CFI/CFF and net
// change in cash, via /calc/cash-flow-statement.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderCashFlowStatement(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.cfs.h1.title">// CASH FLOW STATEMENT</span></h1>
        <p class="muted small" data-i18n="view.cfs.hint.intro">
            The indirect-method cash flow statement reconciles net income to operating cash flow by adding back
            non-cash charges and adjusting for changes in working capital (an increase in a current asset uses
            cash; an increase in a current liability provides cash), then adds the investing and financing
            sections to reach the net change in cash. Drafting aid, not accounting advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.cfs.h2.inputs">Statement inputs</h2>
            <form id="cfs-form" class="inline-form">
                <label><span data-i18n="view.cfs.label.company">Company</span>
                    <input type="text" name="company_name" value="Acme Co" required></label>
                <label><span data-i18n="view.cfs.label.period">Period</span>
                    <input type="text" name="period_label" value="FY2025" required></label>
                <label><span data-i18n="view.cfs.label.ni">Net income ($)</span>
                    <input type="number" step="1000" name="net_income_usd" value="100000" required></label>
                <label><span data-i18n="view.cfs.label.da">Depreciation & amortization ($)</span>
                    <input type="number" step="1000" name="depreciation_amortization_usd" value="20000"></label>
                <label><span data-i18n="view.cfs.label.ar">Change in AR ($, + = increase)</span>
                    <input type="number" step="1000" name="change_in_ar_usd" value="15000"></label>
                <label><span data-i18n="view.cfs.label.inv">Change in inventory ($)</span>
                    <input type="number" step="1000" name="change_in_inventory_usd" value="10000"></label>
                <label><span data-i18n="view.cfs.label.ap">Change in AP ($)</span>
                    <input type="number" step="1000" name="change_in_ap_usd" value="5000"></label>
                <label><span data-i18n="view.cfs.label.capex">Capital expenditures ($)</span>
                    <input type="number" step="1000" name="capex_usd" value="30000"></label>
                <label><span data-i18n="view.cfs.label.sales">Asset sales ($)</span>
                    <input type="number" step="1000" name="asset_sales_usd" value="0"></label>
                <label><span data-i18n="view.cfs.label.debtissued">Debt issued ($)</span>
                    <input type="number" step="1000" name="debt_issued_usd" value="0"></label>
                <label><span data-i18n="view.cfs.label.debtrepaid">Debt repaid ($)</span>
                    <input type="number" step="1000" name="debt_repaid_usd" value="10000"></label>
                <label><span data-i18n="view.cfs.label.equity">Equity issued ($)</span>
                    <input type="number" step="1000" name="equity_issued_usd" value="0"></label>
                <label><span data-i18n="view.cfs.label.dividends">Dividends paid ($)</span>
                    <input type="number" step="1000" name="dividends_usd" value="8000"></label>
                <label><span data-i18n="view.cfs.label.begin">Beginning cash ($)</span>
                    <input type="number" step="1000" name="beginning_cash_usd" value="40000"></label>
                <label><span data-i18n="view.cfs.label.date">Date</span>
                    <input type="date" name="date" value="2026-01-31" required></label>
            </form>
        </div>
        <div id="cfs-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#cfs-form');
    const num = (k) => Number(form.querySelector(`[name="${k}"]`).value) || 0;
    const generate = async () => {
        const body = {
            company_name: (form.querySelector('[name="company_name"]').value || '').trim(),
            period_label: (form.querySelector('[name="period_label"]').value || '').trim(),
            net_income_usd: num('net_income_usd'),
            depreciation_amortization_usd: num('depreciation_amortization_usd'),
            change_in_ar_usd: num('change_in_ar_usd'),
            change_in_inventory_usd: num('change_in_inventory_usd'),
            change_in_ap_usd: num('change_in_ap_usd'),
            capex_usd: num('capex_usd'),
            asset_sales_usd: num('asset_sales_usd'),
            debt_issued_usd: num('debt_issued_usd'),
            debt_repaid_usd: num('debt_repaid_usd'),
            equity_issued_usd: num('equity_issued_usd'),
            dividends_usd: num('dividends_usd'),
            beginning_cash_usd: num('beginning_cash_usd'),
            date: form.querySelector('[name="date"]').value,
        };
        try {
            const doc = await api.calcCashFlowStatement(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.cfs.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase(), ''];
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#cfs-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.cfs.card.cfo">Operating</div>
                    <div class="value">${money(doc.operating_cash_flow_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.cfs.card.cfi">Investing</div>
                    <div class="value">${money(doc.investing_cash_flow_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.cfs.card.cff">Financing</div>
                    <div class="value">${money(doc.financing_cash_flow_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.cfs.card.net">Net change</div>
                    <div class="value">${money(doc.net_change_usd)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.cfs.card.ending">Ending cash</div>
                    <div class="value">${money(doc.ending_cash_usd)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="cfs-copy" type="button" data-i18n="view.cfs.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="cfs-download" type="button" data-i18n="view.cfs.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#cfs-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.cfs.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.cfs.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#cfs-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'cash-flow-statement.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/cash_flow_statement.rs
//! Cash flow statement (indirect method) — reconciles net income to operating
//! cash flow by adding back non-cash charges and adjusting for changes in working
//! capital (an increase in a current asset uses cash; an increase in a current
//! liability provides cash), then adds the investing and financing sections to
//! reach the net change in cash and the ending balance. Distinct from the
//! free-cash-flow module, which derives FCF from a given operating cash flow;
//! this derives operating cash flow itself. Drafting aid, not accounting advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct CashFlowInput {
    pub company_name: String,
    pub period_label: String,
    pub net_income_usd: f64,
    /// Depreciation and amortization (non-cash add-back).
    #[serde(default)]
    pub depreciation_amortization_usd: f64,
    /// Increase (+) or decrease (−) in accounts receivable.
    #[serde(default)]
    pub change_in_ar_usd: f64,
    /// Increase (+) or decrease (−) in inventory.
    #[serde(default)]
    pub change_in_inventory_usd: f64,
    /// Increase (+) or decrease (−) in accounts payable.
    #[serde(default)]
    pub change_in_ap_usd: f64,
    /// Capital expenditures (cash outflow).
    #[serde(default)]
    pub capex_usd: f64,
    /// Proceeds from selling assets.
    #[serde(default)]
    pub asset_sales_usd: f64,
    /// Debt raised.
    #[serde(default)]
    pub debt_issued_usd: f64,
    /// Debt repaid.
    #[serde(default)]
    pub debt_repaid_usd: f64,
    /// Equity raised.
    #[serde(default)]
    pub equity_issued_usd: f64,
    /// Dividends paid.
    #[serde(default)]
    pub dividends_usd: f64,
    /// Beginning cash balance.
    #[serde(default)]
    pub beginning_cash_usd: f64,
    pub date: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct CashFlowStatement {
    pub title: String,
    /// Cash flow from operating activities.
    pub operating_cash_flow_usd: f64,
    /// Cash flow from investing activities.
    pub investing_cash_flow_usd: f64,
    /// Cash flow from financing activities.
    pub financing_cash_flow_usd: f64,
    /// CFO + CFI + CFF.
    pub net_change_usd: f64,
    pub ending_cash_usd: f64,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

/// Format a signed working-capital effect: an increase in an asset is shown as a
/// use of cash (negative).
fn signed(v: f64) -> String {
    if v < 0.0 {
        format!("({})", money(v.abs()))
    } else {
        money(v)
    }
}

pub fn generate(i: &CashFlowInput) -> CashFlowStatement {
    // Increase in a current asset uses cash; increase in a current liability provides cash.
    let cfo = cents(
        i.net_income_usd + i.depreciation_amortization_usd - i.change_in_ar_usd
            - i.change_in_inventory_usd
            + i.change_in_ap_usd,
    );
    let cfi = cents(-i.capex_usd + i.asset_sales_usd);
    let cff = cents(
        i.debt_issued_usd - i.debt_repaid_usd + i.equity_issued_usd - i.dividends_usd,
    );
    let net = cents(cfo + cfi + cff);
    let ending = cents(i.beginning_cash_usd + net);

    let operating_body = format!(
        "Net income {}\n  + Depreciation & amortization {}\n  − Increase in AR {}\n  − Increase in inventory {}\n  + Increase in AP {}\n  = Operating cash flow {}",
        money(i.net_income_usd),
        money(i.depreciation_amortization_usd),
        signed(-i.change_in_ar_usd),
        signed(-i.change_in_inventory_usd),
        signed(i.change_in_ap_usd),
        money(cfo)
    );

    let investing_body = format!(
        "Capital expenditures {}\n  + Asset sales {}\n  = Investing cash flow {}",
        signed(-i.capex_usd),
        money(i.asset_sales_usd),
        money(cfi)
    );

    let financing_body = format!(
        "Debt issued {}\n  − Debt repaid {}\n  + Equity issued {}\n  − Dividends {}\n  = Financing cash flow {}",
        money(i.debt_issued_usd),
        signed(-i.debt_repaid_usd),
        money(i.equity_issued_usd),
        signed(-i.dividends_usd),
        money(cff)
    );

    let clauses = vec![
        DocClause {
            heading: "Statement".into(),
            body: format!("Company: {}\nPeriod: {}\nDate: {}", i.company_name, i.period_label, i.date),
        },
        DocClause { heading: "Operating Activities".into(), body: operating_body },
        DocClause { heading: "Investing Activities".into(), body: investing_body },
        DocClause { heading: "Financing Activities".into(), body: financing_body },
        DocClause {
            heading: "Net Change in Cash".into(),
            body: format!(
                "Net change {} = operating {} + investing {} + financing {}. Beginning cash {} → ending cash {}.",
                money(net), money(cfo), money(cfi), money(cff), money(i.beginning_cash_usd), money(ending)
            ),
        },
    ];

    CashFlowStatement {
        title: "Cash Flow Statement (Indirect Method)".into(),
        operating_cash_flow_usd: cfo,
        investing_cash_flow_usd: cfi,
        financing_cash_flow_usd: cff,
        net_change_usd: net,
        ending_cash_usd: ending,
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> CashFlowInput {
        CashFlowInput {
            company_name: "Acme Co".into(),
            period_label: "FY2025".into(),
            net_income_usd: 100_000.0,
            depreciation_amortization_usd: 20_000.0,
            change_in_ar_usd: 15_000.0,
            change_in_inventory_usd: 10_000.0,
            change_in_ap_usd: 5_000.0,
            capex_usd: 30_000.0,
            asset_sales_usd: 0.0,
            debt_issued_usd: 0.0,
            debt_repaid_usd: 10_000.0,
            equity_issued_usd: 0.0,
            dividends_usd: 8_000.0,
            beginning_cash_usd: 40_000.0,
            date: "2026-01-31".into(),
        }
    }

    #[test]
    fn full_statement() {
        let d = generate(&base());
        assert!(close(d.operating_cash_flow_usd, 100_000.0));
        assert!(close(d.investing_cash_flow_usd, -30_000.0));
        assert!(close(d.financing_cash_flow_usd, -18_000.0));
        assert!(close(d.net_change_usd, 52_000.0));
        assert!(close(d.ending_cash_usd, 92_000.0));
    }

    #[test]
    fn ar_decrease_is_a_source() {
        let d = generate(&CashFlowInput {
            net_income_usd: 50_000.0,
            depreciation_amortization_usd: 5_000.0,
            change_in_ar_usd: -3_000.0, // AR fell → cash collected
            change_in_inventory_usd: 0.0,
            change_in_ap_usd: 0.0,
            capex_usd: 0.0,
            debt_repaid_usd: 0.0,
            dividends_usd: 0.0,
            beginning_cash_usd: 10_000.0,
            ..base()
        });
        assert!(close(d.operating_cash_flow_usd, 58_000.0));
        assert!(close(d.ending_cash_usd, 68_000.0));
    }

    #[test]
    fn sections_sum_to_net_change() {
        let d = generate(&base());
        assert!(close(
            d.operating_cash_flow_usd + d.investing_cash_flow_usd + d.financing_cash_flow_usd,
            d.net_change_usd
        ));
    }

    #[test]
    fn ending_equals_beginning_plus_change() {
        let d = generate(&base());
        assert!(close(d.ending_cash_usd, 40_000.0 + d.net_change_usd));
    }

    #[test]
    fn all_three_sections_present() {
        let d = generate(&base());
        assert!(d.clauses.iter().any(|c| c.heading == "Operating Activities"));
        assert!(d.clauses.iter().any(|c| c.heading == "Investing Activities"));
        assert!(d.clauses.iter().any(|c| c.heading == "Financing Activities"));
    }
}
